tsdb/engine/tsm1/reader.go

```diff
@@ -995,43 +990,35 @@ func (a *indexEntries) Append(entry ...*IndexEntry) {
 	a.entries = append(a.entries, entry...)
 }
 
-func (a *indexEntries) MarshalBinary() (b []byte, err error) {
-	for _, entry := range a.entries {
-		b = append(b, u64tob(uint64(entry.MinTime.UnixNano()))...)
-		b = append(b, u64tob(uint64(entry.MaxTime.UnixNano()))...)
-		b = append(b, u64tob(uint64(entry.Offset))...)
-		b = append(b, u32tob(entry.Size)...)
+func (a *indexEntries) MarshalBinary() ([]byte, error) {
+	buf := make([]byte, len(a.entries)*indexEntrySize)
+	var o int
+
+	for i, entry := range a.entries {
+		o = indexEntrySize * i
+		binary.BigEndian.PutUint64(buf[o:o+8], uint64(entry.MinTime.UnixNano()))
+		binary.BigEndian.PutUint64(buf[o+8:o+16], uint64(entry.MaxTime.UnixNano()))
+		binary.BigEndian.PutUint64(buf[o+16:o+24], uint64(entry.Offset))
+		binary.BigEndian.PutUint32(buf[o+24:o+28], uint32(entry.Size))
 	}
-	return b, nil
+
+	return buf, nil
 }
 
 func (a *indexEntries) Write(w io.Writer) error {
-	for _, entry := range a.entries {
-		_, err := w.Write(u64tob(uint64(entry.MinTime.UnixNano())))
-		if err != nil {
-			return err
-		}
-		_, err = w.Write(u64tob(uint64(entry.MaxTime.UnixNano())))
-		if err != nil {
-			return err
-		}
-
-		_, err = w.Write(u64tob(uint64(entry.Offset)))
-		if err != nil {
-			return err
-		}
-
-		_, err = w.Write(u32tob(entry.Size))
-		if err != nil {
-			return err
-		}
+	// TODO: rename this and change signature to match io.WriterTo
+	// TODO: Possibly split the Marshalling in to smaller chunks (needs analysis first)
+	data, err := a.MarshalBinary()
```

Comment:
Need to revert this and have it write directly to the writer w/o calling `MarshalBinary`.  This was specifically done to avoid large memory spikes and OOMs when writing TSM files.  See ee54a1e.  The issue is that the `indexEntries` slice can be very large so allocating a big slice while multiple are running causes big memory spikes.